src/input_formatter/expression_formatter.ts

```diff
@@ -0,0 +1,155 @@
+		.replace(/(\$)/g, "")
+		.replace(/(\[|\()/g, " bracketLeft ")
+		.replace(/(\]|\))/g, " bracketRight ")
+		.replace(/(<-->|<==>|<=>|<->)/g, " equivalence ")
+		.replace(/(->|-->|=>|==>|\\rightarrow)/g, " implicationRight ")
+		.replace(/(<-|<--|<=|<==|\/leftarrow)/g, " implicationLeft ")
+		.replace(/(neq|not|nicht|¬|-|\\neg)/ig, " negation ")
+		.replace(/(and|und|&|∧|\/\\)/ig, " conjunction ")
+		.replace(/(or|oder|\||∨|\\\/)/ig, " disjunction ")
+		.replace(/(=|==|===|gleich)/g, " equal ")
+		.replace(/(\+|plus|add|sum)/g, " addition ")
+		.replace(/(-|minus)/g, " subtraction ")
+		.replace(/(:|\/|div|geteilt)/g, " division ")
+		.replace(/(\*|mal|mult)/g, " multiplication ")
+		.trim()
+		.replace(/\s{1,}/g, ",");
+	const splittedExpression: string[] = formattedInputExpression.split(/[,]+/g);
+	return splittedExpression;
+
+}
+
+/**
+ * @return finalExpression where exrpession elements got replaced with readable prolog code elements
+ */
+export function replaceExpressionElementsIntoPrologCode(preFormattedExpression: string[]) {
+	const finalFormattedExpression: string[] = [];
+	let index = 0;
+	while (index < preFormattedExpression.length) {
+		if (preFormattedExpression[index].match(allowedExpressionToken)) {
+			const element: string = replaceASingleExpressionElementIntoPrologCode(preFormattedExpression[index]);
+			finalFormattedExpression.push(element);
+		} else {
+			finalFormattedExpression.push(preFormattedExpression[index]);
+		}
+		index++;
+	}
+	return finalFormattedExpression;
+
+}
+
+/**
+ * helper function for replaceExpressionElementsIntoPrologCode
+ * @return finalExpressionElement
+ */
+export function replaceASingleExpressionElementIntoPrologCode(preformattedExpressionElement: string) {
+	const finalExpression: string = preformattedExpressionElement;
+	return finalExpression
+		.replace(/(bracketLeft)/g, "\[")
+		.replace(/(bracketRight)/g, "\]")
+		.replace(/(equivalence)/g, "<->")
+		.replace(/(implicationRight)/g, "->")
+		.replace(/(implicationLeft)/g, "<-")
+		.replace(/(negation)/g, "neg")
+		.replace(/(conjunction)/g, "and")
+		.replace(/(disjunction)/g, "or")
+		.replace(/(equal)/g, "=")
+		.replace(/(addition)/g, "+")
+		.replace(/(subtraction)/g, "-")
+		.replace(/(multiplication)/g, "*")
+		.replace(/(division)/g, ":")
+		;
+}
+
+export function expressionIssueDetector(preFormattedExpression: string[]) {
+	detectBracketIssues(preFormattedExpression);
+	detectMissingStatementsOrConnector(preFormattedExpression);
+}
+
+/**
+ * Check the expression for an equal amount of opened and closed brackets.; it can detect wrong input
+ */
+export function detectBracketIssues(formattedExpression: string[]) {
+	let bracketCount = 0;
+	let index = 0;
+	while (index < formattedExpression.length && bracketCount >= 0) {
+
+		if (formattedExpression[index].match(/(\[|bracketLeft)/)) {
+			bracketCount++;
+		} else if (formattedExpression[index].match(/(\]|bracketRight)/)) {
+			bracketCount--;
+		}
+		index++;
+	}
```

Comment:
i dont know how to use a for loop here when i use a second condition. btw why i should try to avoid while loops? just because to safe some lines of code?

Reply:
For-loops have multiple advantages over while-loops:
* they combine declaring the running-variable, initializing it, incrementing it and checking the loop-condition all in one line, in a form, that is known to all programmers. Makes it easier to read.
* they are super-descriptive about what they are doing. They say "I am looping starting from here until there". That makes it easier to a reader to understand what your code actually does.
* they make it harder to get stuck in an infinite loop. With while-loops, the *change* to the loop-condition is happening somewhere within the body. If that body has multiple branches, it's difficult so always make sure your condition actually gets changed and to maintain it's correctness.

For similar reasons, `foreach`-loops are superior to for-loops:
* they don't declare any more variables than they need to. They also hide the complexity of actually iterating your collection completely away.
* they are super-short and point out your intentions as clearly as possible. "I am iterating this collections elements". 
* It's about impossible to build a non-terminating foreach-loop, unless you build an infinite iterable (please don't). 

I don't differentiate between a foreach-loop and the foreach-function here, and I have no strong opinion on what to use. Please just be consistent, and don't be the first to insert a foreach-loop in a function which only contained foreach-function-calls until then.

Reply:
thx for ur advices. i finally changed this loop into the for each loop deleting the second break condition